File: Java and C++- Client and Server using TCP/Server/src/main/java/bgu/spl/net/impl/stomp/StompFrame.java
package bgu.spl.net.impl.stomp;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentHashMap;
import java.util.HashMap;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentHashMap;

public abstract class StompFrame {

    private String type;
    protected Map<String, String> headers;
    private String body = "";

    public StompFrame(String type1,Map<String, String> headers1, String body){

        this.type=type1;
        this.headers = new LinkedHashMap<>();
        for(String key : headers1.keySet())
        {
            this.headers.put(key, headers1.get(key));
        }
        this.body=body;
    }

    public String getBody() {
        return body;
    }

    public void setBody(String newbody){
        if(newbody!=null)
        this.body=newbody;
    }

    public void setHeaders(Map<String, String> headers1)
    {
        this.headers = new LinkedHashMap<>();
        for(String key : headers1.keySet())
        {
            this.headers.put(key, headers1.get(key));
        }
;
    }

    public String getType() {
        return type;
    }

    public Map<String, String> getHashMap() {
        return headers;
    }


    @Override
    public String toString() {
        String answer = this.type + "\n";
        for (String key : headers.keySet()) {
            answer += key + ":" + headers.get(key) + "\n";
        }
        answer += '\n';
        answer += this.getBody() + "\u0000";
        return answer;
    }
}

File: System_programming/Java and C++- Client and Server using TCP/Client/src/StompFrame.cpp
//
// Created by reut on 11/01/2020.
//

#include "../include/StompFrame.h"
#include <map>
#include <iostream>

using std::string;

StompFrame::StompFrame(string getype, std::unordered_map<string, string> getheaders): type(getype),headers(getheaders),body(" "){}
StompFrame::StompFrame(string getype, std::unordered_map<string, string> getheaders, string body):type(getype),headers(getheaders),body(body) {}

const std::unordered_map<string, string> &StompFrame::getHeaders() const {
    return headers;
}

const string &StompFrame::getBody() const {
    return body;
}

const string &StompFrame::getType() const {
    return type;
}


void StompFrame::setType(const string &type) {
    StompFrame::type = type;
}

void StompFrame::setBody(const string &body) {
    StompFrame::body = body;
}

void StompFrame::setHeaders(const unordered_map<string, string> &headers) {
    this->headers=headers;
}

StompFrame::StompFrame() {

}

string StompFrame::toString() {

    string out;
    string answer;
    map<string, string>::iterator it;
    for (std::pair<std::string, string> element : headers){
        out=out+element.first+":"+element.second+'\n';
    }


    if(body.empty()) {
        answer = type+'\n'+out+'\n';
    }
    else{
        answer = type+'\n'+out+'\n'+body+'\n';
    }

    return answer;
}





